modules/src/events.rs

```diff
@@ -0,0 +1,250 @@
+//      "message.action":["connection_open_init","create_client","connection_open_init"],
+//      "message.module":["ibc_connection","ibc_client",ibc_connection"],
+//      "message.sender":["cosmos187xxg4yfkypl05cqylucezpjvycj24nurvm8p9","cosmos187xxg4yfkypl05cqylucezpjvycj24nurvm8p9","cosmos187xxg4yfkypl05cqylucezpjvycj24nurvm8p9","cosmos187xxg4yfkypl05cqylucezpjvycj24nurvm8p9"],"tm.event":["Tx"],"transfer.amount":["5000stake"],"transfer.recipient":["cosmos17xpfvakm2amg962yls6f84z3kell8c5lserqta"],"tx.hash":["A9E18AE3909F22232F8DBDB1C48F2FECB260A308A2D157E8832E901D45950605"],
+//      "tx.height":["35"]
+// and returns:
+// [{"connection_open_init", 0}, {"create_client", 0}, {"connection_open_init", 1}]
+// where the number in each entry is the index in the matching events that should be used to build the event
+// e.g. for the last "connection_open_init" in the result
+//      "connection_open_init.client_id" -> "testclientsec"
+//      "connection_open_init.connection_id" -> "ancaconnonetestsec",
+//      "connection_open_init.counterparty_client_id" -> "testclientsec","testclientsecsec",
+fn extract_helper(events: &HashMap<String, Vec<String>>) -> Result<Vec<(String, u32)>, String> {
+    let message_action = events.get("message.action").ok_or("Incorrect Event Type")?;
+    let mut val_indeces = HashMap::new();
+    let mut result = Vec::new();
+
+    for action_string in message_action {
+        let idx = val_indeces
+            .entry(action_string.clone())
+            .or_insert_with(|| 0);
+        result.push((action_string.clone(), *idx));
+        *val_indeces.get_mut(action_string.as_str()).unwrap() += 1;
+    }
+    Ok(result)
+}
+
+pub fn get_all_events(result: ResultEvent) -> Result<Vec<IBCEvent>, String> {
+    let mut vals: Vec<IBCEvent> = vec![];
+
+    match &result.data {
+        TMEventData::EventDataNewBlock(nb) => {
+            let block = (&nb.block).as_ref().ok_or("missing block")?;
+            vals.push(NewBlock::new(block.header.height).into());
+        }
+
+        TMEventData::EventDataTx(_tx) => {
+            let events = &result.events.ok_or("missing events")?;
+            let height = events.get("tx.height").ok_or("tx.height")?[0]
+                .parse::<u64>()
+                .map_err(|e| e.to_string())?;
+            let actions_and_indices = extract_helper(&events)?;
+            for action in actions_and_indices {
+                let ev = build_event(RawObject::new(
+                    height.into(),
+                    action.0,
+                    action.1.try_into().unwrap(),
+                    events.clone(),
+                ))
+                .map_err(|e| e.to_string())?;
+                vals.push(ev);
+            }
+        }
+        _ => {}
+    }
+    Ok(vals)
+}
+
+pub fn build_event(object: RawObject) -> Result<IBCEvent, BoxError> {
+    match object.action.as_str() {
+        "create_client" => Ok(IBCEvent::from(ClientEvents::CreateClient::try_from(
```

Comment:
I am wondering if action string can somehow be defined as part of the corresponding IBCEvent so we don't hard code it here.

Reply:
https://doc.rust-lang.org/edition-guide/rust-2018/trait-system/associated-constants.html

Reply:
I thought about it a bit but couldn't find quickly a good solution. The other problem was that these I was not easy to find these strings so this was a sandbox for figuring this out. I will raise another PR to finish the packet events and cleanup the code.